### DISCUSSION TITLE
Application Insights Dependency tracking with TCP protocol/DirectMode

### DISCUSSION THREAD

**Discussion by fernandozpiccin** [Azure/azure-cosmos-dotnet-v3#3446] [Q&A]
I'm just wondering how clients are implementing Cosmos dependency tracking in Application Insights (AI).
We are using TCP mode and the out-of-box dependency tracking isn't supported.
We have been trying to implement the dependency using a custom `RequestHandler`. We have taken similar approach as done in the [AI SDK for Document DB](https://github.com/microsoft/ApplicationInsights-dotnet/blob/main/WEB/Src/DependencyCollector/DependencyCollector/Implementation/HttpParsers/DocumentDbHttpParser.cs) where it inspects the Cosmos URIs and determines the operation name.
However, this approach isn't working with Cosmos SDK V3 as the operation names aren't been precisely mapped on our custom request handler. For example, when creating a new document/item using `CreateItemStreamAsync` the Cosmos SDK V3 is building the request as `POST /dbs/myDatabase/colls/myCollection` instead of `POST /dbs/myDatabase/colls/myCollection/docs` which makes our mapping (same DocumentDB implementaion) to result in an operation name as "Query collection" instead of "Create Document" or "Query Document".
Looking the `TelemetryHandler`, I can see that it contains all the information we would be looking for in terms of tracking Cosmos as dependency. It's a shame that we can't take advantage of that.
Another alternative to build the correct operation name is having access to the `ResourceType` and `OperationType` properties of `RequestMessage` - at least as `public get` as they are currently defined with  `internal` protection level.

I'd appreciate if anyone could assist with the following questions, please:

1. Why the `TelemetryHandler` can't be executed/configured form clients using TCP/Direct mode?
2. Is there any intention of making the RequestMessage to expose a bit more of the operation information so that custom handlers like the one we tried to build can have access and do a little more with the information in context e.g. having access to properties like `ResourceType` and `OperationType` or `ResponseMessage.Headers.SubStatusCodeLiteral`? 
3. Why is the Cosmos SDK V3 building requests to create document items without including the `/docs` segment? It is creating the request URI as `POST /dbs/myDatabase/colls/myCollection` instead `POST /dbs/myDatabase/colls/myCollection/docs`. Looking into the [Cosmos REST API endpoints](https://docs.microsoft.com/en-us/rest/api/cosmos-db/collections), it doesn't mention the `POST /dbs/myDatabase/colls/myCollection` as an endpoint.

Hope to hear from the SDK team about this.
Many thanks.

**Comment by sourabh1007** [comment]
ref. https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/sdk-observability?tabs=dotnet

FYI: Cosmos db SDK and Cosmos REST API are 2 separate libraries. Here, we are talking about, only cosmos db SDK.
